provider.go

```diff
@@ -15,102 +15,267 @@
-			RegistryClientOpts: co,
-			RootCerts:          fulcio.GetRoots(),
-		})
+		metadata := &checkedMetadata{}
 
+		checkedSignatures, bundleVerified, err := cosign.VerifyImageSignatures(ctx, ref, co)
 		if err != nil {
-			fmt.Println(err)
-			sendResponse(nil, fmt.Sprintf("VerifyImageSignatures: %v", err), w)
-			return
+			return nil, fmt.Errorf("verifyImageSignatures: %v", err)
 		}
 
-		if bundleVerified {
-			fmt.Println("signature verified for:", key)
-			fmt.Printf("%d number of valid signatures found for %s, found signatures: %v\n", len(checkedSignatures), key, checkedSignatures)
-			results = append(results, externaldata.Item{
-				Key:   key,
-				Value: key + "_valid",
-			})
-		} else {
-			fmt.Printf("no valid signatures found for: %s\n", key)
-			results = append(results, externaldata.Item{
-				Key:   key,
-				Error: key + "_invalid",
-			})
+		if co.RekorClient != nil && !bundleVerified {
+			return nil, fmt.Errorf("no valid signatures found for %s: %v", key, err)
 		}
+
+		if len(checkedSignatures) == 0 {
+			return nil, fmt.Errorf("no valid signatures found for %s", key)
+		}
+
+		metadata.ImageSignatures, err = formatSignaturePayloads(checkedSignatures)
+		if err != nil {
+			return nil, fmt.Errorf("formatSignaturePayloads: %v", err)
+		}
+
+		fmt.Println("signature verified for: ", key)
+		fmt.Printf("%d number of valid signatures found for %s, found signatures: %v\n", len(checkedSignatures), key, checkedSignatures)
+
+		if o.AttestationPresent {
+			fmt.Println("Verifying Attestations for image: ", key)
+
+			checkedAttestations, bundleVerified, err := cosign.VerifyImageAttestations(ctx, ref, co)
+			if err != nil {
+				return nil, fmt.Errorf("verifyImageAttestations: %v", err)
+			}
+			if co.RekorClient != nil && !bundleVerified {
+				return nil, fmt.Errorf("no valid attestations found for: %s", key)
+			}
+
+			AttestationPayloads, err := formatAttestations(checkedAttestations)
+			if err != nil {
+				return nil, fmt.Errorf("formatAttestations: %v", err)
+			}
+
+			metadata.AttestationSignatures = AttestationPayloads
+
+			fmt.Println("attestation verified for: ", key)
+			fmt.Printf("%d number of valid attestations found for %s, found attestations: %v\n", len(checkedAttestations), key, checkedAttestations)
+		}
+
+		return metadata, nil
+	}
+
+	return nil, fmt.Errorf("no verifier found for: %s", key)
+}
+
+// formatAttestations takes the payload within an Attestation and base64 decodes it, returning it as an in-toto statement
+func formatAttestations(verifiedAttestations []oci.Signature) ([]in_toto.Statement, error) {
+
+	decodedAttestations := make([]in_toto.Statement, len(verifiedAttestations))
+
+	for i, att := range verifiedAttestations {
+		p, err := att.Payload()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "error fetching payload: %v", err)
+			return nil, err
+		}
+
+		var pm map[string]interface{}
+		json.Unmarshal(p, &pm)
+
+		payload := strings.Trim(fmt.Sprintf("%v", pm["payload"]), "\"")
```

Comment:
I'm not very happy with this line. It seemed that when I got the payload on line [213](https://github.com/sigstore/cosign-gatekeeper-provider/pull/8/files#diff-1bb2cdf24b7aa1b1986f79d4ea823ec921a95ba94f6436119cb2a74c21ebff82R213), I still needed to get the payload field within it and remove the `"` fields around the base64 encoded value.

@ribbybibby @developer-guy @dlorenc feel free to input if you have any ideas